tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -342,6 +380,265 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
+                        clamp_upper);
+    }
+    return OkStatus();
+  }
+
+  Status F8Scaled(HloInstruction *instr, HloInstruction *a, HloInstruction *b,
+                  HloInstruction *a_scale, HloInstruction *b_scale) {
+    // FP8 GEMM kernels are only available on Hopper and newer architectures.
+    if (!cuda_compute_capability_.IsAtLeast(
+            se::CudaComputeCapability::HOPPER)) {
+      return OkStatus();
+    }
+
+    // cuBLASLt FP8 GEMM kernels require one of the two operands to be in
+    // F8E4M3FN format.
+    if (!((a->shape().element_type() == F8E4M3FN &&
+           b->shape().element_type() == F8E4M3FN) ||
+          (a->shape().element_type() == F8E4M3FN &&
+           b->shape().element_type() == F8E5M2) ||
+          (a->shape().element_type() == F8E5M2 &&
+           b->shape().element_type() == F8E4M3FN))) {
+      return OkStatus();
+    }
+
+    // cuBLASLt FP8 GEMM kernels require the operand sizes to be multiples
+    // of 16.
+    for (int i = 0; i < a->shape().dimensions_size(); ++i) {
+      if (a->shape().dimensions(i) % 16) {
+        return OkStatus();
+      }
+    }
+    for (int i = 0; i < b->shape().dimensions_size(); ++i) {
+      if (b->shape().dimensions(i) % 16) {
+        return OkStatus();
+      }
+    }
+
+    // cuBLASLt FP8 GEMM kernels require the scaling factors to be in F32
+    // format.
+    HloInstruction *a_scale_f32 =
+        instr->AddInstruction(HloInstruction::CreateConvert(
+            ShapeUtil::MakeScalarShape(F32), a_scale));
+    HloInstruction *b_scale_f32 =
+        instr->AddInstruction(HloInstruction::CreateConvert(
+            ShapeUtil::MakeScalarShape(F32), b_scale));
+
+    PrimitiveType c_type;
+    switch (instr->shape().element_type()) {
+      case F8E4M3FN:
+      case F8E5M2:
+      case BF16:
+        c_type = BF16;
+        break;
+      case F16:
+        c_type = F16;
+        break;
+      case F32:
+        c_type = F32;
+        break;
+      default:
+        return OkStatus();
+    }
+
+    // TODO(philipphack): Consider enabling the epilogue fusion of a matrix bias
+    // for FP8 GEMMs.
+    Literal c_literal = LiteralUtil::Zero(c_type);
+    HloInstruction *c = instr->AddInstruction(
+        HloInstruction::CreateConstant(c_literal.Clone()));
+    HloInstruction *c_bcast = instr->AddInstruction(
+        HloInstruction::CreateBroadcast(instr->shape(), c, {}));
+
+    Literal one_literal = LiteralUtil::One(F32);
+    HloInstruction *one = instr->AddInstruction(
+        HloInstruction::CreateConstant(one_literal.Clone()));
+
+    // cuBLASLt FP8 GEMM kernels require A, which is later exchanged with B, to
+    // be transposed.
+    HloInstruction *b_transp =
+        instr->AddInstruction(HloInstruction::CreateTranspose(
+            ShapeUtil::MakeShape(
+                b->shape().element_type(),
+                {b->shape().dimensions(1), b->shape().dimensions(0)}),
+            b, {1, 0}));
```

Comment:
Instead of transposing here, can we adjust the layout in matmul_utils.cc, swapping between `CUBLASLT_ORDER_COL` and `CUBLASLT_ORDER_ROW`?

I'm a bit confused why cublasLT exposes both a per-input `cublasLtOrder_t` to specify between row-order and column-order and also exposes a per-input `cublasOperation_t` to specify whether the input should be transposed. Isn't a row-order nontransposed input equivalent to a column-order tranposed input?

Also, it's not guaranteed the A and B will be swapped. I think they are only swapped if the output is in row-order.

Reply:
I thought so too but adjusting the layout isn’t sufficient, and the left operand must be actually transposed. In other words, a transpose operation and a change of the layout are not equivalent from the perspective of the library.

Reply:
Out of curiosity, @philipphack, how are they nonequivalent from the perspective of the library? Does the documentation explain this difference?

Reply:
I can't speak to what happens inside the library. I think the documentation states that A has to be transposed without relating it to the layout.

Reply:
Asked internally.

Reply:
(I'm one of cuBLAS devs @pjannaty asked to respond here.)

I'll check if the documentation needs to be updated to be more precise: the TN in the fp8 docs refers to all matrices in column-major layout (so C and D are non-transposed).

The `cublasLtOrder_t` is provided for convenience. Actual matrix layout is figured out internally by also taking `cublasOperation_t` into account. So you can construct multiple valid combinations that all will work as long as they map to column-major TN.

Reply:
Thanks for the info!

Somewhat unrelated, but the cuBLAS documentation for CUDA 12.0 seemed to have removed the info describing what combination of FP8 dtypes are supported. In particular, Table 3 in the [11.8 documentation](https://docs.nvidia.com/cuda/archive/11.8.0/cublas/index.html#cublasLtMatmul) had the FP8 combinations, but Table 3 in the [12.0 documentation](https://docs.nvidia.com/cuda/cublas/index.html#id9) instead describes the complex number combinations, which was Table 4 in the 11.8 documentation. Do you know why the FP8 info was removed?

Reply:
(used the wrong account in the previous response)

> Do you know why the FP8 info was removed?

This was not intentional. I will let our docs team know.

Reply:
Documentation has been fixed: https://docs.nvidia.com/cuda/cublas/index.html#cublasltmatmul.